### PR TITLE
py-igraph: new package with version 0.10.6

### PR DESCRIPTION
**Pull request by ChristopherChristofi**

### CODE REVIEW

**Comment by szhorvat** [comment]
Great to see the latest python-igraph in Spack!

Consider making explicit choices about whether vendored or external versions of libraries would be used. If this is not done, igraph will use external versions when present, and vendored ones otherwise. In other words, the result is unpredictable.

You can take a look at how I did this in MacPorts:

https://github.com/macports/macports-ports/blob/master/python/py-igraph/Portfile

Basically, set the `IGRAPH_CMAKE_EXTRA_ARGS` envvar, which controls the configuration of the igraph C library. Docs for that are at https://igraph.org/c/html/latest/igraph-Installation.html

This is only relevant when python-igraph is built with its bundled copy of the igraph C library.

----

Another option is to link to the igraph C library that is already in Spack. You will need to update it to 0.10.6 as well first.

**Comment by ChristopherChristofi** [comment]
Thank you @szhorvat for the pointers.

Ran into some difficulty when trying to directly link py-igraph to an existing igraph build:
```
192      /usr/bin/ld: /tmp/_igraph.abi3.so.gUw8IL.ltrans31.ltrans.o: relocation R_X86_64_TPOFF32 against hidden symbol `temp_count.lto_priv.0' can not be used when making a shared object
193      /usr/bin/ld: /tmp/_igraph.abi3.so.gUw8IL.ltrans32.ltrans.o: relocation R_X86_64_TPOFF32 against hidden symbol `igraph_i_error_handler.lto_priv.0' can not be used when making a shared object
194      /usr/bin/ld: /tmp/_igraph.abi3.so.gUw8IL.ltrans33.ltrans.o: relocation R_X86_64_TPOFF32 against hidden symbol `igraph_i_finally_stack_size' can not be used when making a shared object
195      /usr/bin/ld: /tmp/_igraph.abi3.so.gUw8IL.ltrans34.ltrans.o: relocation R_X86_64_TPOFF32 against hidden symbol `igraph_i_error_handler.lto_priv.0' can not be used when making a shared object
196      /usr/bin/ld: /tmp/_igraph.abi3.so.gUw8IL.ltrans36.ltrans.o: relocation R_X86_64_TPOFF32 against hidden symbol `igraph_i_rng_default' can not be used when making a shared object
197      /usr/bin/ld: /tmp/_igraph.abi3.so.gUw8IL.ltrans37.ltrans.o: relocation R_X86_64_TPOFF32 against hidden symbol `igraph_i_finally_stack_size' can not be used when making a shared object
  >> 198    collect2: error: ld returned 1 exit status
```

Separately, I came across another issue when trying to locate (external) blas lapack for py-igraph, there must be something I am missing when it comes to passing arguments to cmake through the inbuilt python build system:
```
>> 746    /usr/bin/ld: cannot find -lblas
>> 747      /usr/bin/ld: cannot find -llapack
>> 748      collect2: error: ld returned 1 exit status
```

So I opted for using the vendored versions of blas and lapack. I hope this resolves some of the original unpredictability.

**Comment by szhorvat** [comment]
If you link against an external igraph, make sure it was built as a shared library. Linking against an external static library is going to be fussy.

File: var/spack/repos/builtin/packages/igraph/package.py [hunk 1]
```diff
@@ -6,12 +6,42 @@
...
+        conditional("autotools", when="@0.7.1"),
+        default="cmake",
+    )
+
+    with when("build_system=cmake"):
+        depends_on("arpack-ng", when="@0.10.6")
+        depends_on("blas", when="@0.10.6")
+        depends_on("glpk+gmp@4.57:", when="@0.10.6")
+        depends_on("gmp", when="@0.10.6")
+        depends_on("lapack", when="@0.10.6")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/igraph/package.py]: `when="@0.10.6"` isn't needed, `build_system=cmake` implies `@0.10.6`

File: var/spack/repos/builtin/packages/igraph/package.py [hunk 2]
```diff
@@ -6,12 +6,42 @@
...
 
     homepage = "https://igraph.org/"
     url = "https://github.com/igraph/igraph/releases/download/0.7.1/igraph-0.7.1.tar.gz"
 
+    version("0.10.6", sha256="99bf91ee90febeeb9a201f3e0c1d323c09214f0b5f37a4290dc3b63f52839d6d")
     version("0.7.1", sha256="d978030e27369bf698f3816ab70aa9141e9baf81c56cc4f55efbe5489b46b0df")
 
+    build_system(
+        conditional("cmake", when="@0.10.6"),
+        conditional("autotools", when="@0.7.1"),
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/igraph/package.py]: Would be nice to know what range of versions support which build systems. Makes it easier to add new versions in the future.

File: var/spack/repos/builtin/packages/igraph/package.py [hunk 3]
```diff
@@ -6,12 +6,42 @@
...
+    with when("build_system=cmake"):
+        depends_on("arpack-ng", when="@0.10.6")
+        depends_on("blas", when="@0.10.6")
+        depends_on("glpk+gmp@4.57:", when="@0.10.6")
+        depends_on("gmp", when="@0.10.6")
+        depends_on("lapack", when="@0.10.6")
+
     depends_on("libxml2")
+
+    def cmake_args(self):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/igraph/package.py:33]: It looks like Autotools also has similar flags you can add

File: var/spack/repos/builtin/packages/py-igraph/package.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    variant("matplotlib", default=False, description="Enable plotting with Matplotlib")
+
+    depends_on("arpack-ng")
+    depends_on("cmake")
+    depends_on("gmp")
+    depends_on("glpk+gmp@4.57:")
+    depends_on("libxml2")
+    depends_on("py-setuptools", type=("build", "run"))
+    depends_on("py-texttable@1.6.2:", type=("build", "run"))
+    depends_on("py-matplotlib@3.5.0:3.6.0", type=("run"), when="+matplotlib")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-igraph/package.py]:
```suggestion
    depends_on("py-matplotlib@3.5", type="run", when="+matplotlib")
```
Spack version ranges are inclusive, 3.6.0 shouldn't be included

File: var/spack/repos/builtin/packages/py-igraph/package.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    homepage = "https://github.com/igraph/python-igraph"
+    pypi = "igraph/igraph-0.10.6.tar.gz"
+
+    version("0.10.6", sha256="76f7aad294514412f835366a7d9a9c1e7a34c3e6ef0a6c3a1a835234323228e8")
+
+    variant("matplotlib", default=False, description="Enable plotting with Matplotlib")
+
+    depends_on("arpack-ng")
+    depends_on("cmake")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-igraph/package.py]: Probably build-only

**Review by adamjstewart — COMMENTED** [review]
Ideally we always want to avoid vendored dependencies

File: var/spack/repos/builtin/packages/igraph/package.py (hunk 2 above)
Comment (szhorvat) [reply var/spack/repos/builtin/packages/igraph/package.py]: 0.9 and later use CMake.

**Comment by szhorvat** [comment]
Note that matplotlib is a not a hard dependency for python-igraph, and there's probably no reason to specify it as such. Some feature will only work when matplotlib os present, but that same can be said for a number of other soft dependencies such as numpy, scipy, networkx, graph-tool, etc.

The only Python package that is a hard dependency is texttable.

**Comment by ChristopherChristofi** [comment]
> Note that matplotlib is a not a hard dependency for python-igraph, and there's probably no reason to specify it as such.

Thank you, yes with reference to your requirements listed in the setup I have treated matplotlib as a soft-dependency, it has only been added as a non-default variant for runtime specifically - I assume enabling this feature at the variant level is simpler for spack users when compared to multiple explicit installs to replicate the same environment

File: var/spack/repos/builtin/packages/igraph/package.py [hunk 6]
```diff
@@ -6,12 +6,49 @@
...
     url = "https://github.com/igraph/igraph/releases/download/0.7.1/igraph-0.7.1.tar.gz"
 
+    version("0.10.6", sha256="99bf91ee90febeeb9a201f3e0c1d323c09214f0b5f37a4290dc3b63f52839d6d")
     version("0.7.1", sha256="d978030e27369bf698f3816ab70aa9141e9baf81c56cc4f55efbe5489b46b0df")
 
+    variant("shared", default=False, description="Enable shared build")
+
+    build_system(
+        conditional("cmake", when="@0.9:"),
+        conditional("autotools", when="@:0.9"),
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/igraph/package.py]:
```suggestion
        conditional("autotools", when="@:0.8"),
```
Version ranges in Spack are inclusive. So 0.8 is the most recent version that supports Autotools

File: var/spack/repos/builtin/packages/py-igraph/package.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    pypi = "igraph/igraph-0.10.6.tar.gz"
+
+    version("0.10.6", sha256="76f7aad294514412f835366a7d9a9c1e7a34c3e6ef0a6c3a1a835234323228e8")
+
+    variant("matplotlib", default=False, description="Enable plotting with Matplotlib")
+
+    depends_on("cmake", type="build")
+    depends_on("igraph+shared@0.10.6", when="@0.10.6")
+    depends_on("pkgconfig", type="build")
+    depends_on("py-setuptools", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-igraph/package.py]:
```suggestion
    depends_on("py-setuptools", type="build")
```

**Review by adamjstewart — DISMISSED** [review]

**Comment by adamjstewart** [comment]
@spackbot fix style

**Comment by adamjstewart** [comment]
spackbot doesn't have permission to push to your branch but you can run it yourself or run `spack style --fix` and it will automatically fix the style issues.

**Comment by ChristopherChristofi** [comment]
@spackbot fix style

**Review by adamjstewart — APPROVED** [review]
